scripts: west: runners:  Add device type argument for hostname devices

Martinhoff-maker opened the pull request:
The goal of this PR is to add the support of a new argument ("--dev-id-type") for silabs_commander runner and Jlink runner. It will allow to flash device through an hostname.

pdgendt commented:
Not against this, but this runner supports `--tool-opt` to pass options to executable underneath.

Martinhoff-maker commented:
Yes I know but It would be more explicit this way for a new user I think. At the start, I wish to mimic Jlink script (to be able to pass the IP directly in the device argument) but I would also like to use hostname as IP argument. So I ended with the IP argument then I don't have to parse the arg.

mbolivar reviewed (commented):
This runner already supports `--tool-opt` (`-O`). Why do we need yet-another option when `west flash -O` can already supply `--ip`?

mbolivar commented:
Sorry, I see @pdgent already asked this

jerome-pouiller commented:
So, the idea would be:
```
  --dev-id-type={auto,ip,serialno,uart} Specify if --dev-id is an IP, a serial port or a
                                        serial number (as returned by the USB host). 
  --ip=ADDR                             Shortcut for "--dev-id-type=ip --dev-id=ADDR"
  --uart=DEVICE                         Shortcut for "--dev-id-type=tty --dev-id=DEVICE"
  --serialno=SERIAL                     Shortcut for "--dev-id-type=serialno --dev-id=SERIAL"
```

Then, `--dev-id-type=auto` could behave as follow:
   - dev-id matches with `[0-9]+` -> `serialno`
   - dev-id is an existing file  -> `tty`
   - dev-id is resolved by gethostbyname() -> `ip`
   - legacy behavior otherwise

Does it sound good?

Martinhoff-maker committed:
script: west_commands: jlink: add device type for hostname dev

This commit adds automatic type detection for the dev-id argument and
introduces a device-type option to explicitly tell JLink how to handle
the dev-id argument. It enables hostname usage for flashing and
debugging.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
script: west_commands: silabs: add device type for hostname dev

This commit adds automatic type detection for the dev-id argument and
introduces a device-type option to explicitly tell Silabs Commander how
to handle the dev-id argument. It enables hostname usage for flashing
devices.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

jerome-pouiller commented on scripts/west_commands/runners/jlink.py:
```diff
@@ -464,13 +523,36 @@ def run_flash_cmd(self, fname, flash_file, **kwargs):
+                device_args = ['-USB', f'{self.dev_id}']
+                self.logger.warning(f'Could not auto-detect device type for "{self.dev_id}",'
+                                  f' using USB as fallback. Use --dev-id-type to specify'
+                                  f' explicitly.')
```
I am not sure this message is meaningful for user. In addition, JLink will probably also fail. I suggest a brief message: `f'Could not auto-detect device type for "{self.dev_id}"'`. Or `f'"{self.dev_id}" does not match any known pattern'`.

jhedberg commented:
Are there any changes needed to the `scripts/requirements*.txt` files, or are those new python modules already covered by existing requirements?
